File: src/main/java/org/tindalos/principle/domain/resultprocessing/reporter/DesignQualityCheckResultsReporter.java
package org.tindalos.principle.domain.resultprocessing.reporter;

import java.util.Map;

import org.tindalos.principle.domain.checker.DesignQualityCheckResults;
import org.tindalos.principle.domain.coredetector.CheckResult;
import org.tindalos.principle.domain.coredetector.ViolationsReporter;

import com.google.common.collect.Maps;

public class DesignQualityCheckResultsReporter {

	private Map<Class<? extends CheckResult>, ViolationsReporter<? extends CheckResult>> reporters;

	public DesignQualityCheckResultsReporter( Map<Class<? extends CheckResult>, ViolationsReporter<? extends CheckResult>> reporters) {
		this.reporters = Maps.newHashMap(reporters);
	}
	
	public void report(DesignQualityCheckResults results, Printer printer) {

		for (CheckResult checkResult : results.resultList()) {
			ViolationsReporter reporter = reporters.get(checkResult.getClass());
			String report = reporter.report(checkResult);

			if (checkResult.violationsDetected()) {
				printer.printWarning(report);
			} else {
				printer.printInfo(report);
			}
		}
	}
	
	public void setReporters(Map<Class<? extends CheckResult>, ViolationsReporter<? extends CheckResult>> reporters) {
		this.reporters = Maps.newHashMap(reporters);
	}

}

File: src/main/java/org/tindalos/principle/domain/resultprocessing/reporter/DesignQualityCheckResultsReporter.scala
package org.tindalos.principle.domain.resultprocessing.reporter

import org.tindalos.principle.domain.checker.DesignQualityCheckResults
import org.tindalos.principle.domain.coredetector.CheckResult
import org.tindalos.principle.domain.detector.acd.ACDResult
import org.tindalos.principle.domain.detector.adp.ADPResult
import org.tindalos.principle.domain.detector.layering.LayerViolationsResult
import org.tindalos.principle.domain.detector.sap.SAPResult
import org.tindalos.principle.domain.detector.sdp.SDPResult
import org.tindalos.principle.domain.detector.submodulesblueprint.SubmodulesBlueprintCheckResult
import org.tindalos.principle.domain.detector.thirdparty.ThirdPartyViolationsResult

//couldn't figure out how to inject reporter functions in a Map
object DesignQualityCheckResultsReporter {

  def buildResultReporter(reportAdpResult: ADPResult => String,
                          reportLayerViolationsResult: LayerViolationsResult => String,
                          reportThirdPartyViolationsResult: ThirdPartyViolationsResult => String,
                          reportSAPResult: SAPResult => String,
                          reportACDResult: ACDResult => String,
                          reportSubmodulesBlueprintCheckResult: SubmodulesBlueprintCheckResult => String,
                          reportSDPResult: SDPResult => String ) = {

    def getReport(checkResult: CheckResult) = {
      val report = checkResult match {
        case cr: ADPResult => reportAdpResult(cr)
        case cr: LayerViolationsResult => reportLayerViolationsResult(cr)
        case cr: ThirdPartyViolationsResult => reportThirdPartyViolationsResult(cr)
        case cr: SDPResult => reportSDPResult(cr)
        case cr: SAPResult => reportSAPResult(cr)
        case cr: ACDResult => reportACDResult(cr)
        case cr: SubmodulesBlueprintCheckResult => reportSubmodulesBlueprintCheckResult(cr)
        case _ => throw new RuntimeException("terrible thing - no result type")
      }
      (report, checkResult.expectationsFailed())
    }

    (results: DesignQualityCheckResults)
    => for (checkResult <- results.checkResults) yield getReport(checkResult)
  }


}